### PR TITLE
Add Privacy API - foundation of privacy awareness

### PR DESCRIPTION
**Pull request by ayomawdb**
## Introduction

General Data Protection Regulation (GDPR) has became a relatively hard problem to address, because most of the programs or integrations are not done with privacy and transparency requirements in focus. It is infeasible to provide a solution to implicitly makes a program 100% GDPR compliant, but we can provide easy access to libraries required to handle most of the tasks programmers have to manually perform. In order to do so, it is required to make Ballerina a natively Privacy and GDPR aware programming language.

One requirement in being GDPR compliant is that Personally Identifiable Information (PII) should be pseudonymised before being stored. Therefore, PII remains in a one storage location, separated from other non-PII data. If it is required to delete PII or comply with an information request, it is now possible to easily isolate PII and act on those.

Currently, programmers should do all the required custom coding to handle pseudonymisation which is complex and error-prone.  

## Solution 

This PR will introduce a `Privacy API` to Ballerina standard library. `Privacy API` acts as the foundation of making Ballerina programming language Privacy /  GDPR aware. This API is capable of operating on a `PII Store` by allowing pseudonymise, depseudonymise and delete operations. Programmers can configure a `PII Store` and invoke the `Privacy API` as follows to offload pseudonymisation related logic to the stdlib.

```ballerina
import ballerina/h2;
import ballerina/privacy;

endpoint h2:Client testDB {
    path: "./target/H2PIIStore/",
    name: "TestDBH2",
    username: "SA",
    password: "",
    poolOptions: { maximumPoolSize: 1 }
};

@final string TABLE_NAME = "PII_STORE";
@final string ID_CLOUMN = "id";
@final string PII_COLUMN = "pii";

public function main (string... args) {
    privacy:H2PiiStore piiStore = new(testDB, TABLE_NAME, ID_CLOUMN, PII_COLUMN);

    var pseudonymizeResult = privacy:pseudonymize(piiStore, pii);
    // pseudonymizeResult is string|error
    // string is the pseudonymized identifier 

    // Example pseudonymized identifier  (UUID)
    string id = "37288f57-00db-4dc6-89b2-f5afd5c43fe6";

    var depseudonymizeResult = privacy:depseudonymize(piiStore, id);
    // depseudonymizeResult is string|error
    // string is the depseudonymized PII value 

    var deleteResult = privacy:delete(piiStore, id);
    // deleteResult is error?
}
```
Currently, any database supported by Ballerina can be used as `PII Store`. There will be external packages that extend the capabilities of the `Privacy API` by introducing support for different storage mechanisms such as:

- AWS Secret Manager 
- Vault by HashiCorp

It is possible to write custom PII storage mechanisms and plug them into the `Privacy API` just by implementing an object similar to the [PiiStore abstract object](https://github.com/ballerina-platform/ballerina-lang/compare/master...ayomawdb:privacy-api?expand=1#diff-d80f20ee07109a28e7c419b824630b03).

In future, this `Privacy API` will be utilized by other libraries within Ballerina to provide easier and transparent handling of PII. 

Relevant further details are discussed in mail thread: [Making Ballerina GDPR Aware - Approach](https://groups.google.com/forum/#!topic/ballerina-dev/XNlLdfbAUSc) and the public document: [Ballerina for Privacy / GDPR Aware Programming](https://docs.google.com/document/d/1vJjs7Xm4_-wLj8zLo4WlrjgaXcaP4odFQjWXOaFRpOs).

### CODE REVIEW

File: stdlib/privacy/src/main/ballerina/privacy/database_pii_store_utils.bal [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+# Represents personally identifiable information (PII) stored in a database
+#
+# + pii - personally identifiable information
+type PiiData record {
+    string pii;
```
Comment (wggihan) [inline stdlib/privacy/src/main/ballerina/privacy/database_pii_store_utils.bal]: Don't we need to seal this record?

File: stdlib/privacy/src/main/ballerina/privacy/database_pii_store_utils.bal [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+};
+
+# Build insert query based on the table name and column names
+#
+# + tableName - table name used to store PII
+# + idColumn - column name used to store pseudonymized identifier
+# + piiColumn - column name used to store PII
+# + return - insert query
+function buildInsertQuery (string tableName, string idColumn, string piiColumn) returns string {
+    return "INSERT INTO " + tableName + " (" + idColumn + "," + piiColumn + ") VALUES (?, ?)";
```
Comment (wggihan) [inline stdlib/privacy/src/main/ballerina/privacy/database_pii_store_utils.bal]: We have io:sprintf function. Can we utilize that in here? WDYT?

File: stdlib/privacy/src/main/ballerina/privacy/h2_pii_store.bal [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    # Pseudonymize personally identifiable information (PII) and store PII and the pseudonymized identifier
+    #
+    # + pii - PII to be pseudonymized
+    # + return - 36 characters long UUID if storage operation was successful, error if storage operation failed
+    public function pseudonymize (string pii) returns string|error {
+        endpoint h2:Client client = clientEndpoint;
+        string dbQuery = buildInsertQuery(tableName, idColumn, piiColumn);
+        string id = system:uuid();
+        sql:Parameter paramId = {sqlType: sql:TYPE_VARCHAR, value: id};
+        sql:Parameter paramPii = {sqlType: sql:TYPE_VARCHAR, value: pii};
```
Comment (anupama-pathirage) [inline stdlib/privacy/src/main/ballerina/privacy/h2_pii_store.bal]: No need to sql:Parameters here. The id and pii variables can be directly passed into update action.

File: stdlib/privacy/src/main/ballerina/privacy/h2_pii_store.bal [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    }
+
+    # Depseudonymize the identifier by retrieving the personally identifiable information (PII)
+    #
+    # + id - pseudonymized identifier to be depseudonymize
+    # + return - PII if retrieval was successful, error if retrieval failed
+    public function depseudonymize (string id) returns string|error {
+        endpoint h2:Client client = clientEndpoint;
+        string dbQuery = buildSelectQuery(tableName, idColumn, piiColumn);
+        sql:Parameter paramId = {sqlType: sql:TYPE_VARCHAR, value: id};
```
Comment (anupama-pathirage) [inline stdlib/privacy/src/main/ballerina/privacy/h2_pii_store.bal]: Remove parameter and directly pass the id to the action. Check all places. We can directly pass primitive types to sql actions as varargs.

File: stdlib/privacy/src/main/ballerina/privacy/database_pii_store_utils.bal (hunk 1 above)
Comment (ayomawdb) [reply stdlib/privacy/src/main/ballerina/privacy/database_pii_store_utils.bal]: +1. Will update.

File: stdlib/privacy/src/main/ballerina/privacy/database_pii_store_utils.bal (hunk 2 above)
Comment (ayomawdb) [reply stdlib/privacy/src/main/ballerina/privacy/database_pii_store_utils.bal]: +1. Will update.

File: stdlib/privacy/src/main/ballerina/privacy/jdbc_pii_store.bal [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+import ballerina/system;
+
+# Represents personally identifiable information (PII) storage mechanisum based on JDBC compatible database
+#
+# + clientEndpoint - reference to H2 database client endpoint
+# + tableName - table name used to store PII
+# + idColumn - column name used to store pseudonymized identifier
+# + piiColumn - column name used to store PII
+public type JdbcPiiStore object {
+    public jdbc:Client clientEndpoint;
```
[the remark this answers is not in the document]
Comment (ayomawdb) [reply stdlib/privacy/src/main/ballerina/privacy/jdbc_pii_store.bal]: Tried this. If this is made public, we cannot use abstract-objects to make the "store" implementation pluggable. "JdbcPiiStore" cannot be assigned to a "PiiStore", where there are private fields.

File: stdlib/privacy/src/main/ballerina/privacy/h2_pii_store.bal (hunk 3 above)
Comment (ayomawdb) [reply stdlib/privacy/src/main/ballerina/privacy/h2_pii_store.bal]: +1. Will update.

File: stdlib/privacy/src/main/ballerina/privacy/h2_pii_store.bal (hunk 4 above)
Comment (ayomawdb) [reply stdlib/privacy/src/main/ballerina/privacy/h2_pii_store.bal]: +1. Will update.

**Pull request merged**